File: Base/크루스칼/크루스칼.py
import sys
sys.stdin = open("크루스칼.txt", 'r')

V, E = map(int, input().split())
# parent 만들기
parent = [i for i in range(V+1)]

# 비용순으로 정렬된 간선 만들기
edges = []
for _ in range(E):
    a, b, cost = map(int, input().split())
    edges.append((cost, a, b))
edges.sort()
result = 0

# 유니온 파인드


def find(parent, x):
    if parent[x] != x:
        parent[x] = find(parent, parent[x])
    return parent[x]


def union(parent, a, b):
    a = find(parent, a)
    b = find(parent, b)
    if a > b:
        parent[b] = a
    else:
        parent[a] = b

# 크루스칼


for edge in edges:
    cost, a, b = edge
    # 루트 노드가 같지 않으면 집합에 포함(사이클이 발생하지 않는 경우)
    if find(parent, a) != find(parent, b):
        union(parent, a, b)
        result += cost
print(result)

File: BASE/JSCategory/트리/크루스칼/크루스칼.js
var fs = require("fs");
// var input = fs.readFileSync("/dev/stdin", "utf8").split("\n");
var input = fs.readFileSync("./1922.txt", "utf8").split("\r\n");
var idx = 0;
const V = parseInt(input[idx++]);
const E = parseInt(input[idx++]);

var parent = new Array(V + 1).fill(0).map((_, i) => i);
var edges = [];
for (let i = 0; i < E; i++) {
  [a, b, cost] = input[idx++].split(" ").map(Number);
  edges.push([cost, a, b]);
}
edges.sort((a, b) => a[0] - b[0]);
let result = 0;
const find = (x) => {
  if (parent[x] === x) return x;
  return (parent[x] = find(parent[x]));
};

const union = (_a, _b) => {
  var a = find(_a);
  var b = find(_b);
  parent[a] = b;
};

for (let [cost, a, b] of edges) {
  if (find(a) !== find(b)) {
    union(a, b);
    result += cost;
  }
}
console.log(result);
